Public/Src/Engine/Scheduler/Scheduler.cs

```diff
@@ -5458,22 +5458,26 @@ string IFileContentManagerHost.GetProducerDescription(in FileOrDirectoryArtifact
         }
 
         /// <summary>
-        /// Gets the first consumer description associated with a FileOrDirectory artifact.
+        /// Gets the consumer description if and only if a single consumer is associated with a FileOrDirectory artifact.
         /// </summary>
         /// <param name="artifact">The artifact for which to get the first consumer description.</param>
         /// <returns>The first consumer description or null if there is no consumer.</returns>
         [SuppressMessage("Microsoft.Design", "CA1033:InterfaceMethodsShouldBeCallableByChildTypes")]
         public string GetConsumerDescription(in FileOrDirectoryArtifact artifact)
         {
             var producerId = PipGraph.GetProducer(artifact);
+            string singleConsumer = null;
             foreach (var consumerEdge in DataflowGraph.GetOutgoingEdges(producerId.ToNodeId()))
             {
+                if (singleConsumer != null) {
+                    return null;    // More than 1 consumers
+                }
                 Pip consumer = PipGraph.GetPipFromPipId(consumerEdge.OtherNode.ToPipId());
-                return consumer.GetDescription(Context);
+                singleConsumer = consumer.GetDescription(Context);
             }
 
-            // No consumer
-            return null;
+            // Single consumer description or null for 0 consumers
+            return singleConsumer;

```

Comment:
I don't know if this is the right change, but if it is, can you please write it in a less convoluted way, e.g.,:
```csharp
    var producerId = PipGraph.GetProducer(artifact);
    var numConsumers = DataflowGraph.GetOutgoingEdgesCount(producerId.ToNodeId());
    if (numConsumers == 1)
    {
        var consumerPipId = DataflowGraph.GetOutgoingEdges(producerId.ToNodeId()).First().OtherNode.ToPipId();
        var consumer = PipGraph.GetPipFromPipId(consumerPipId);
        return consumer.GetDescription(Context);
    }
    else
    {
        return null;
    }
```
No `for` loop is needed if we only care about the first consumer and/or whether there exist more than 1 consumer.